File: Session4_Strings/substring-concat.java
class Solution {
    /**
     * Disclaimer: Not the most efficient solution to find
     * the answer, but it's a good way to start thinking about
     * Sliding windows
     */
    public List<Integer> findSubstring(String s, String[] words) {
        /* Find number of words */
        int nWord = words.length;
        /* Number of letters per word */
        int nLetter = words[0].length();
        /* Required sliding window */
        int window = nWord * nLetter;
        /* Answer storing the starting indexes */
        List ans = new ArrayList<Integer>();
        /* Hashmap equivalent of String[] words (parameter) */
        Map<String, Integer> map1 = new HashMap<>();

        /* Populate map with word count */
        for (String word : words) {
            map1.put(word, map1.getOrDefault(word, 0) + 1);
        }

        /* Notice the for-loop constraints to prevent Out-of-Bounds */
        for (int i = 0; i < s.length() - window + 1; i++) {
            /* HashMap to record all words that's inside */
            Map<String, Integer> map2 = new HashMap<>();
            /* For-loop to populate map2 with word counts */
            for (int j = i; j < i + window; j += nLetter) {
                String fragment = s.substring(j, j + nLetter);
                map2.put(fragment, map2.getOrDefault(fragment, 0) + 1);
            }
            /* If word count equal to previous map, add to answer */
            if (map1.equals(map2))
                ans.add(i);
        }

        return ans;
    }
}

File: Session4_Strings/substring-concat.py
class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        """
        Disclaimer: Not the most efficient way to do this question,
        But a good way to start thinking about sliding windows.
        """
        # Number of words and number of letters per word
        nWord, nLetter = len(words), len(words[0])
        # Length of sliding window
        window = nWord * nLetter
        # output
        ans = []
        # Notice the for-loop conditions to prevent out-of-bounnds
        for i in range(0, len(s) - window + 1):
            # copy array
            temp = words.copy()
            # range of sliding window, step size set at nLetter
            for j in range(i, i + window, nLetter):
                # Collect substring
                fragment = s[j:j + nLetter]
                # Remove fragment from temp if prevent
                if fragment in temp:
                    temp.remove(fragment)
            # If empty, means that all the strings in words are 
            # contained in that particular substring. Record starting index
            if temp == []:
                ans.append(i)
        return ans